indexer/test/advertisement-walker.test.js

```diff
@@ -0,0 +1,411 @@
+      tail: undefined, // previous walk was finished
+      lastHead: knownAdvertisement.adCid, // arbitrary advertisement
+      status: 'some-status'
+    }
+
+    const { newState, finished } = await processNextAdvertisement({ providerId, providerInfo, walkerState })
+
+    assert.deepStrictEqual(newState, /** @type {WalkerState} */({
+      head: undefined, // we finished the walk, there is no head
+      tail: undefined, // we finished the walk, there is no next step
+      lastHead: FRISBII_AD_CID, // lastHead was updated to head of the walk we finished
+      status: `All advertisements from ${newState?.lastHead} to the end of the chain were processed.`
+    }))
+
+    assert.strictEqual(finished, true, 'finished')
+  })
+
+  it('updates lastHead after tail reaches lastHead', async () => {
+    /** @type {ProviderInfo} */
+    const providerInfo = {
+      providerAddress,
+      lastAdvertisementCID: knownAdvertisement.adCid
+    }
+
+    /** @type {WalkerState} */
+    const walkerState = {
+      head: knownAdvertisement.adCid,
+      tail: knownAdvertisement.previousAdCid,
+      lastHead: knownAdvertisement.previousPreviousAdCid,
+      status: 'some-status'
+    }
+
+    const { newState, indexEntry, finished } = await processNextAdvertisement({ providerId, providerInfo, walkerState })
+
+    assert.deepStrictEqual(newState, /** @type {WalkerState} */({
+      head: undefined, // we finished the walk, there is no head
+      tail: undefined, // we finished the walk, there is no next step
+      lastHead: walkerState.head, // lastHead was updated to head of the walk we finished
+      status: `All advertisements from ${newState?.lastHead} to the end of the chain were processed.`
+    }))
+
+    assert(indexEntry, 'the step found an index entry')
+
+    assert.strictEqual(finished, true, 'finished')
+  })
+
+  it('handles Fetch errors and explains the problem in the status', async () => {
+    /** @type {ProviderInfo} */
+    const providerInfo = {
+      providerAddress: 'http://127.0.0.1:80/',
+      lastAdvertisementCID: 'baguqeeraTEST'
+    }
+
+    const result = await processNextAdvertisement({ providerId, providerInfo, walkerState: undefined })
+
+    assert.deepStrictEqual(result, {
+      failed: true,
+      newState: {
+        head: 'baguqeeraTEST',
+        tail: 'baguqeeraTEST',
+        lastHead: undefined,
+        status: 'Error processing baguqeeraTEST: HTTP request to http://127.0.0.1/ipni/v1/ad/baguqeeraTEST failed: connect ECONNREFUSED 127.0.0.1:80'
+      }
+    })
+  })
+
+  it('handles timeout errors and explains the problem in the status', async () => {
+    /** @type {ProviderInfo} */
+    const providerInfo = {
+      providerAddress: 'http://127.0.0.1:80/',
+      lastAdvertisementCID: 'baguqeeraTEST'
+    }
+
+    const result = await processNextAdvertisement({
+      providerId,
+      providerInfo,
+      walkerState: undefined,
+      fetchTimeout: 1
+    })
```

Comment:
There is a race condition in this test: sometimes, we get back the HTTP response before the timeout fires. I'll fix it soon.

Reply:
Fixed in 1e31450